File: GOOGLE SOLUTION CHALLANGE/sportshield/backend/app/routes/assets.py
import uuid
from datetime import datetime
from typing import Optional, List
from fastapi import APIRouter, Depends, UploadFile, File, Form, HTTPException
from pydantic import BaseModel
from app.config.firebase import assets_ref, transactions_ref, users_ref, ArrayUnion, safe_update
from app.routes.auth import get_current_user, require_role
from app.services.storage_service import storage_service
from app.services.blockchain_service import blockchain_service
from app.services.ai_service import ai_service

router = APIRouter(prefix="/assets", tags=["assets"])

class AssetCreate(BaseModel):
    title: str
    description: Optional[str] = ""
    media_type: str
    sports_category: Optional[str] = "general"
    event: Optional[str] = ""
    royalty_percentage: Optional[int] = 10
    price: Optional[float] = 0.0

class PurchaseRequest(BaseModel):
    asset_id: str
    buyer_wallet: str

@router.post("/upload-asset")
async def upload_asset(
    title: str = Form(...),
    description: str = Form(""),
    media_type: str = Form(...),
    sports_category: str = Form("general"),
    event: str = Form(""),
    royalty_percentage: int = Form(10),
    price: float = Form(0.0),
    file: UploadFile = File(...),
    user: dict = Depends(get_current_user),
):
    """Upload asset, store in Firebase, generate hash, mint NFT."""
    file_bytes = await file.read()
    if len(file_bytes) == 0:
        raise HTTPException(status_code=400, detail="Empty file")

    # Upload to Firebase Storage
    content_type = file.content_type or "application/octet-stream"
    media_url = storage_service.upload_file(file_bytes, file.filename, content_type)

    # Generate hashes
    sha256_hash = storage_service.compute_sha256(file_bytes)
    phash = ai_service.compute_perceptual_hash(file_bytes) if media_type.startswith("image") else ""

    # Run fraud detection baseline
    fraud_result = ai_service.detect_fraud(file_bytes, media_type)

    # Get user info
    user_doc = users_ref.document(user["uid"]).get()
    user_data = user_doc.to_dict() or {}
    wallet_address = user_data.get("wallet_address", "")

    # Create asset ID
    asset_id = f"asset_{datetime.utcnow().strftime('%Y%m%d%H%M%S')}_{uuid.uuid4().hex[:8]}"

    # Mint on blockchain
    token_price_wei = int(price * 1e18)  # Convert MATIC to wei
    blockchain_result = blockchain_service.register_asset(
        owner_address=wallet_address or user["uid"],
        token_uri=media_url,
        metadata={
            "title": title,
            "description": description,
            "mediaType": media_type,
            "sportsCategory": sports_category,
            "eventName": event,
            "location": "",
            "recordedAt": int(datetime.utcnow().timestamp()),
            "metadataHash": sha256_hash,
            "c2paCertificateId": f"c2pa_{asset_id}",
        },
        royalty_percentage=royalty_percentage,
        price_wei=token_price_wei,
    )

    # C2PA-inspired metadata layer
    c2pa_manifest = {
        "claim": "AthletiChain Protected",
        "timestamp": datetime.utcnow().isoformat(),
        "algorithm": "SHA-256",
        "signature": "Simulated-C2PA-Signature",
        "signer": "AthletiChain CA",
        "creator": user["uid"],
        "content_hash": sha256_hash,
    }

    # Save to Firestore
    asset_doc = {
        "asset_id": asset_id,
        "title": title,
        "description": description,
        "media_type": media_type,
        "sports_category": sports_category,
        "event": event,
        "owner_id": user["uid"],
        "owner_wallet": wallet_address,
        "price": price,
        "royalty_percentage": royalty_percentage,
        "status": "active" if not fraud_result["fraud_detected"] else "flagged",
        "media_url": media_url,
        "content_hash": sha256_hash,
        "perceptual_hash": phash,
        "c2pa_manifest": c2pa_manifest,
        "blockchain": {
            "token_id": blockchain_result.get("tokenId", "0"),
            "contract_address": blockchain_result.get("contractAddress", ""),
            "transaction_hash": blockchain_result.get("txHash", ""),
            "network": "polygon_mumbai",
        },
        "ai_detection": {
            "baseline_fraud_detected": fraud_result["fraud_detected"],
            "baseline_confidence": fraud_result["confidence_score"],
            "deepfake_probability": fraud_result["deepfake_probability"],
            "analyzed_at": datetime.utcnow().isoformat(),
        },
        "ownership_history": [
            {
                "action": "minted",
                "from": "0x0000...0000",
                "to": wallet_address or user["uid"],
                "date": datetime.utcnow().isoformat(),
            }
        ],
        "created_at": datetime.utcnow().isoformat(),
        "updated_at": datetime.utcnow().isoformat(),
    }
    assets_ref.document(asset_id).set(asset_doc)

    # Update user stats
    current_count = user_data.get("stats", {}).get("assets_count", 0)
    users_ref.document(user["uid"]).update({"stats.assets_count": current_count + 1})

    return {
        "success": True,
        "asset_id": asset_id,
        "token_id": blockchain_result.get("tokenId", "0"),
        "blockchain_tx_id": blockchain_result.get("txHash", ""),
        "fraud_check": fraud_result,
    }

@router.post("/detect-fraud")
async def detect_fraud(
    file: UploadFile = File(...),
    media_type: str = Form("image"),
    user: dict = Depends(get_current_user),
):
    """Detect fraud for an uploaded file."""
    file_bytes = await file.read()
    result = ai_service.detect_fraud(file_bytes, media_type)
    return {"success": True, **result}

@router.post("/buy-asset")
async def buy_asset(req: PurchaseRequest, user: dict = Depends(get_current_user)):
    """Purchase an asset."""
    asset_doc = assets_ref.document(req.asset_id).get()
    if not asset_doc.exists:
        raise HTTPException(status_code=404, detail="Asset not found")

    asset_data = asset_doc.to_dict()
    if asset_data.get("status") != "active":
        raise HTTPException(status_code=400, detail="Asset not available for purchase")

    seller_id = asset_data["owner_id"]
    price = float(asset_data.get("price", 0))
    royalty_pct = int(asset_data.get("royalty_percentage", 10))

    # Calculate royalty split
    royalty_amount = price * (royalty_pct / 100)
    platform_amount = royalty_amount * 0.1
    athlete_amount = royalty_amount - platform_amount
    creator_amount = price - royalty_amount

    # Update ownership in Firestore
    buyer_doc = users_ref.document(user["uid"]).get()
    buyer_data = buyer_doc.to_dict() or {}

    safe_update(assets_ref.document(req.asset_id), {
        "owner_id": user["uid"],
        "owner_wallet": req.buyer_wallet,
        "status": "sold",
        "updated_at": datetime.utcnow().isoformat(),
        "ownership_history": ArrayUnion([{
            "action": "sold",
            "from": asset_data.get("owner_wallet", seller_id),
            "to": req.buyer_wallet,
            "date": datetime.utcnow().isoformat(),
        }]),
    })

    # Record transaction
    tx_id = f"tx_{datetime.utcnow().strftime('%Y%m%d%H%M%S')}_{uuid.uuid4().hex[:6]}"
    transactions_ref.document(tx_id).set({
        "transaction_id": tx_id,
        "type": "purchase",
        "asset_id": req.asset_id,
        "buyer_id": user["uid"],
        "buyer_wallet": req.buyer_wallet,
        "seller_id": seller_id,
        "seller_wallet": asset_data.get("owner_wallet", ""),
        "amount": price,
        "royalty_split": {
            "athlete": round(athlete_amount, 2),
            "creator": round(creator_amount, 2),
            "platform": round(platform_amount, 2),
        },
        "status": "completed",
        "created_at": datetime.utcnow().isoformat(),
    })

    return {
        "success": True,
        "transaction_id": tx_id,
        "asset_id": req.asset_id,
        "new_owner": req.buyer_wallet,
        "royalty_distributed": {
            "athlete": f"{athlete_amount:.2f} MATIC",
            "creator": f"{creator_amount:.2f} MATIC",
            "platform": f"{platform_amount:.2f} MATIC",
        },
    }

@router.get("/")
async def list_assets(
    category: Optional[str] = None,
    min_price: Optional[float] = None,
    max_price: Optional[float] = None,
    sort_by: Optional[str] = "newest",
    page: int = 1,
    limit: int = 20,
):
    """List marketplace assets."""
    query = assets_ref.where("status", "in", ["active", "sold"])
    if category:
        query = query.where("sports_category", "==", category)

    docs = query.stream()
    assets = []
    for doc in docs:
        data = doc.to_dict()
        data["id"] = doc.id
        assets.append(data)

    # Price filtering (client-side for simplicity)
    if min_price is not None:
        assets = [a for a in assets if float(a.get("price", 0)) >= min_price]
    if max_price is not None:
        assets = [a for a in assets if float(a.get("price", 0)) <= max_price]

    # Sorting
    if sort_by == "price_asc":
        assets.sort(key=lambda x: float(x.get("price", 0)))
    elif sort_by == "price_desc":
        assets.sort(key=lambda x: float(x.get("price", 0)), reverse=True)
    else:  # newest
        assets.sort(key=lambda x: x.get("created_at", ""), reverse=True)

    total = len(assets)
    start = (page - 1) * limit
    paginated = assets[start:start + limit]

    return {
        "success": True,
        "assets": paginated,
        "pagination": {
            "page": page,
            "limit": limit,
            "total": total,
            "pages": (total + limit - 1) // limit,
        },
    }

@router.get("/{asset_id}")
async def get_asset(asset_id: str, user: dict = Depends(get_current_user)):
    """Get asset details."""
    doc = assets_ref.document(asset_id).get()
    if not doc.exists:
        raise HTTPException(status_code=404, detail="Asset not found")
    data = doc.to_dict()
    data["id"] = doc.id
    return {"success": True, "asset": data}

@router.post("/freeze-asset")
async def freeze_asset(
    asset_id: str = Form(...),
    reason: str = Form(...),
    user: dict = Depends(get_current_user),
):
    """Admin: freeze a suspicious asset."""
    require_role(user, ["admin"])

    doc = assets_ref.document(asset_id).get()
    if not doc.exists:
        raise HTTPException(status_code=404, detail="Asset not found")

    data = doc.to_dict()
    token_id = int(data.get("blockchain", {}).get("token_id", 0))

    # Call blockchain freeze
    blockchain_service.freeze_asset(token_id, reason)

    assets_ref.document(asset_id).update({
        "status": "frozen",
        "freeze_reason": reason,
        "frozen_at": datetime.utcnow().isoformat(),
        "updated_at": datetime.utcnow().isoformat(),
    })

    return {"success": True, "message": "Asset frozen successfully"}

File: GOOGLE SOLUTION CHALLANGE/sportshield/backend/src/routes/assets.ts
// filepath: backend/src/routes/assets.ts
import { Router, Request, Response } from 'express';
import { body, validationResult } from 'express-validator';
import { authMiddleware } from '../middleware/auth';
import { assetService } from '../services/assetService';
import { logger } from '../utils/logger';

const router = Router();

// Validation rules
const registerAssetValidation = [
  body('title').notEmpty().trim().isLength({ max: 200 }),
  body('description').optional().trim().isLength({ max: 1000 }),
  body('media_type').isIn(['image', 'video', 'audio']),
  body('media_url').isURL(),
  body('sports_category').optional().trim(),
  body('event').optional().trim(),
  body('royalty_percentage').optional().isInt({ min: 0, max: 100 }),
  body('price').optional().isNumeric()
];

const verifyContentValidation = [
  body('asset_id').notEmpty()
];

const detectFraudValidation = [
  body('media_url').isURL(),
  body('media_type').isIn(['image', 'video'])
];

const purchaseAssetValidation = [
  body('asset_id').notEmpty(),
  body('buyer_wallet').matches(/^0x[a-fA-F0-9]{40}$/)
];

/**
 * POST /api/v1/assets/register
 * Register a new digital asset
 */
router.post('/register', 
  authMiddleware,
  registerAssetValidation,
  async (req: Request, res: Response) => {
    try {
      const errors = validationResult(req);
      if (!errors.isEmpty()) {
        return res.status(400).json({ 
          success: false, 
          error: { code: 'VALIDATION_ERROR', details: errors.array() }
        });
      }

      const userId = req.user?.uid;
      const assetData = req.body;

      const result = await assetService.registerAsset(userId!, assetData);

      res.status(201).json({
        success: true,
        asset_id: result.assetId,
        blockchain_tx_id: result.blockchainTxId,
        token_id: result.tokenId,
        c2pa_certificate_id: result.c2paCertificateId,
        message: 'Asset registered successfully'
      });
    } catch (error: any) {
      logger.error('Asset registration error:', error);
      res.status(500).json({ 
        success: false, 
        error: { code: 'REGISTRATION_FAILED', message: error.message }
      });
    }
  }
);

/**
 * POST /api/v1/assets/verify
 * Verify content authenticity
 */
router.post('/verify', 
  authMiddleware,
  verifyContentValidation,
  async (req: Request, res: Response) => {
    try {
      const { asset_id } = req.body;

      const result = await assetService.verifyAsset(asset_id);

      res.json({
        success: true,
        is_authentic: result.isAuthentic,
        verification_details: result.verificationDetails,
        owner: result.owner,
        verified_at: result.verifiedAt
      });
    } catch (error: any) {
      logger.error('Asset verification error:', error);
      res.status(500).json({ 
        success: false, 
        error: { code: 'VERIFICATION_FAILED', message: error.message }
      });
    }
  }
);

/**
 * POST /api/v1/assets/detect-fraud
 * Detect potential fraud or unauthorized content
 */
router.post('/detect-fraud', 
  authMiddleware,
  detectFraudValidation,
  async (req: Request, res: Response) => {
    try {
      const { media_url, media_type } = req.body;

      const result = await assetService.detectFraud(media_url, media_type);

      res.json({
        success: true,
        fraud_detected: result.fraudDetected,
        confidence_score: result.confidenceScore,
        similar_assets: result.similarAssets,
        deepfake_probability: result.deepfakeProbability,
        recommendation: result.recommendation
      });
    } catch (error: any) {
      logger.error('Fraud detection error:', error);
      res.status(500).json({ 
        success: false, 
        error: { code: 'DETECTION_FAILED', message: error.message }
      });
    }
  }
);

/**
 * POST /api/v1/assets/purchase
 * Purchase a digital asset
 */
router.post('/purchase', 
  authMiddleware,
  purchaseAssetValidation,
  async (req: Request, res: Response) => {
    try {
      const { asset_id, buyer_wallet } = req.body;
      const buyerId = req.user?.uid;

      const result = await assetService.purchaseAsset(asset_id, buyerId!, buyer_wallet);

      res.json({
        success: true,
        transaction_id: result.transactionId,
        transaction_hash: result.transactionHash,
        asset_id: result.assetId,
        new_owner: result.newOwner,
        royalty_distributed: result.royaltyDistributed,
        purchased_at: result.purchasedAt
      });
    } catch (error: any) {
      logger.error('Asset purchase error:', error);
      res.status(500).json({ 
        success: false, 
        error: { code: 'PURCHASE_FAILED', message: error.message }
      });
    }
  }
);

/**
 * GET /api/v1/assets/:id
 * Get asset details
 */
router.get('/:id', 
  authMiddleware,
  async (req: Request, res: Response) => {
    try {
      const { id } = req.params;

      const asset = await assetService.getAsset(id);

      if (!asset) {
        return res.status(404).json({ 
          success: false, 
          error: { code: 'NOT_FOUND', message: 'Asset not found' }
        });
      }

      res.json({
        success: true,
        asset
      });
    } catch (error: any) {
      logger.error('Get asset error:', error);
      res.status(500).json({ 
        success: false, 
        error: { code: 'GET_FAILED', message: error.message }
      });
    }
  }
);

/**
 * GET /api/v1/assets/ownership/:id
 * Get ownership information
 */
router.get('/ownership/:id', 
  authMiddleware,
  async (req: Request, res: Response) => {
    try {
      const { id } = req.params;

      const ownership = await assetService.getOwnership(id);

      res.json({
        success: true,
        asset_id: id,
        ownership,
        blockchain: ownership.blockchain
      });
    } catch (error: any) {
      logger.error('Get ownership error:', error);
      res.status(500).json({ 
        success: false, 
        error: { code: 'GET_FAILED', message: error.message }
      });
    }
  }
);

/**
 * GET /api/v1/assets
 * List assets (marketplace)
 */
router.get('/', 
  async (req: Request, res: Response) => {
    try {
      const { 
        category, 
        min_price, 
        max_price, 
        sort_by, 
        page = '1', 
        limit = '20' 
      } = req.query;

      const result = await assetService.listAssets({
        category: category as string,
        minPrice: min_price as string,
        maxPrice: max_price as string,
        sortBy: sort_by as string,
        page: parseInt(page as string),
        limit: parseInt(limit as string)
      });

      res.json({
        success: true,
        assets: result.assets,
        pagination: result.pagination
      });
    } catch (error: any) {
      logger.error('List assets error:', error);
      res.status(500).json({ 
        success: false, 
        error: { code: 'LIST_FAILED', message: error.message }
      });
    }
  }
);

export default router;